tls/s2n_ktls.c

```diff
@@ -117,6 +203,18 @@ static S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktl
     return S2N_RESULT_OK;
 }
 
+S2N_RESULT s2n_ktls_configure_crypto(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_GUARD(s2n_ktls_validate(conn, ktls_mode));
+
```

Comment:
Why are you validating a second time? If you want to do the validation closer to where values are used, do any crypto / cipher suite validation in s2n_ktls_set_keys instead of in s2n_ktls_validate.

Reply:
I know we try to keep things minimal but ktls_validate is a good sanity check that should always be true in this file. I have added this to all functions so that we have a nice invariant precaution before executing anything. Since this is not performant code I feel the invariant is good to have for all fn in this file.

This also helps with testing incase we test individual fns and havnt setup the connection correctly.

Reply:
Sure, this isn't performant code and there's no real harm in calling a lightweight function repeatedly. You could even put ktls_validate in a little for loop to make _real_ sure ktls is valid :) But it's unnecessary here. You just need to check once before enabling ktls.

And if there is specific validation that s2n_prf_generate_key_material or s2n_ktls_set_keys rely on and will misbehave without, they should perform that validation themselves rather than assuming that it's performed elsewhere. For example-- does s2n_ktls_set_keys have any check to ensure that the cipher is actually aes128_gcm before it assumes that it is?

Reply:
https://github.com/aws/s2n-tls/pull/4071/files#diff-952fe646831caa6b2a901b8bba3058e4e08e67795e0bb27e4ad0320d2ae783a0R127-R132

I have a TODO comment which states that we need a unique check and how we should do this. However, since we are only adding AES_GCM_128 support to start, I dont think this need to be addressed right now. It should be added before adding support for another cipher though.

Reply:
> you should also hardcode the check for AES_GCM_128

I dont think there is a way to do that currently, since there is no unique identifier for ciphers. Possibly I am missing something?

Reply:
addressed in https://github.com/aws/s2n-tls/pull/4071/commits/7c1d107e195266e15fef64903a54e69c7710cec8